plugins/poll/lib/poll.rb

```diff
@@ -31,43 +45,103 @@ def self.vote(user, post_id, poll_name, options)
+        if poll.irv?
+          # for IRV, we need to remove all votes and re-create them as there is no way to update them due to lack of primary key.
+          PollVote.where(poll: poll, user: user).delete_all
+          creation_set = new_option_ids
+        else
+          # remove non-selected votes
+          PollVote
+            .where(poll: poll, user: user)
+            .where.not(poll_option_id: new_option_ids)
+            .delete_all
+          creation_set = new_option_ids - old_option_ids
+        end
 
         # create missing votes
-        (new_option_ids - old_option_ids).each do |option_id|
-          PollVote.create!(poll: poll, user: user, poll_option_id: option_id)
+        creation_set.each do |option_id|
+          if poll.irv?
+            option_digest = poll.poll_options.find(option_id).digest
+
+            PollVote.create!(
+              poll: poll,
+              user: user,
+              poll_option_id: option_id,
+              rank: options.find { |o| o[:digest] == option_digest }[:rank],
+            )
+          else
+            PollVote.create!(poll: poll, user: user, poll_option_id: option_id)
+          end
         end
       end
 
-    # Ensure consistency here as we do not have a unique index to limit the
-    # number of votes per the poll's configuration.
-    is_multiple = serialized_poll[:type] == "multiple"
-    offset = is_multiple ? (serialized_poll[:max] || serialized_poll[:options].length) : 1
-
-    DB.query(<<~SQL, poll_id: poll_id, user_id: user.id, offset: offset)
-    DELETE FROM poll_votes
-    USING (
-      SELECT
-        poll_id,
-        user_id
-      FROM poll_votes
-      WHERE poll_id = :poll_id
-      AND user_id = :user_id
-      ORDER BY created_at DESC
-      OFFSET :offset
-    ) to_delete_poll_votes
-    WHERE poll_votes.poll_id = to_delete_poll_votes.poll_id
-    AND poll_votes.user_id = to_delete_poll_votes.user_id
-    SQL
+    unless serialized_poll[:type] == IRV
+      # Ensure consistency here as we do not have a unique index to limit the
+      # number of votes per the poll's configuration.
+      is_multiple = serialized_poll[:type] == MULTIPLE
+      offset = is_multiple ? (serialized_poll[:max] || serialized_poll[:options].length) : 1
+
+      params = { poll_id: poll_id, offset: offset, user_id: user.id }
+
+      DB.query(<<~SQL, params)
+      DELETE FROM poll_votes
+      USING (
+        SELECT
+          poll_id,
+          user_id
+        FROM poll_votes
+        WHERE poll_id = :poll_id
+        AND user_id = :user_id
+        ORDER BY created_at DESC
+        OFFSET :offset
+      ) to_delete_poll_votes
+      WHERE poll_votes.poll_id = to_delete_poll_votes.poll_id
+      AND poll_votes.user_id = to_delete_poll_votes.user_id
+      SQL
+    end
+
+    if serialized_poll[:type] == IRV
+      serialized_poll[:irv_outcome] = DiscoursePoll::Irv.irv_outcome(poll_id)
+    end
```

Comment:
Instead of writing `unless foo` and then `if foo` I would just write it as `if foo ... else ...`?

Reply:
@ZogStriP Well, yes, it's a nice tidy up flow wise but logically they are not functionally related directly.

Generating an IRV outcome has nothing to do with maintaining consistency for other vote types.

I'm kind of on the fence here, but ok let's change it because it reads better.

Reply:
Committed this change https://github.com/discourse/discourse/pull/27155/commits/687b412a578443403d8a9d10f6dad5da793fa641